README.md
# ascii-art
Python module for converting images/batches of images to ASCII ART

## Installation
1. Install python 2.7 [link](https://www.python.org/downloads/release/python-2713/)
2. `sudo easy_install pip`
3. `sudo pip install virtualenv`
4. `git clone <this repo>`
5. `cd ascii-art`
6. `virtualenv venv`
7. `. venv/bin/activate`
8. `pip install --editable ascii`
9. `ulimit -n 2000`


changelog.md
# Changelog

## 0.1

## 0.2

1. Output sequences are now PNGs with transparent backgrounds.
2. --help messages added.
3. 'Reduce Flicker' function works. NOTE: reducing flicker will also inherently reduce the variety of characters from which it can choose. ANOTHER NOTE: Cranking the reduce flicker up to something like 127 (roughly half of 255, meaning pixels are basically either considered black or white) can have some neat effects.
4. Characters list can be edited without error.
5. You SHOULDN'T need to run ulimit -n abignumber anymore?
6. Characters list is stored in external 'chars.txt' file. The first line is prepended by a '#',
and is only there to serve as a reference when creating new character sets. To create a new character set,
copy this line without the '#' (NOTE: the line begins with a space) and paste it below, removing/adding whatever you want. You can store multiple character sets by prepending sets you dont want to currently use with a '#'. The program will use the first character set it finds without a '#' at the beginning.
7. A new optional argument '--outfile' has been added. By default, the program will create a folder named like <INFILE NAME> + '_out', and will overwrite this folder each time the given infile is processed. If you want to process the same file multiple times, you can specify --outfile <DESCRIPTIVE FILE NAME>.
8. All arguments (except for --infile and --outfile) have defaults specified in ascii/settings.py. You can edit these so you dont have to specify them manually each time. You can also override select settings by specifying their argument in the command line.


## 0.3

1. Characters are printed in color of pixel sampled to choose them. (To override this provide optional argument nocolor)
2. --infile now takes (and requires) an argument of a video, such as 'coolclip.mov' or 'upsetting.mpeg'


test.py
from ascii import reduce_flicker
from random import randint

frames = [[[randint(1,20) for _ in range(2)] for _ in range(2)] for _ in range(10)]

print "before de jittering"
for frame in frames:
    for row in frame:
        print row
    print

reduce_flicker(frames, 10)

print "after de jittering"
for frame in frames:
    for row in frame:
        print row
    print


edge.py
from PIL import Image, ImageEnhance, ImageDraw, ImageFont
# import numpy as np
# from matplotlib import pyplot as plt

CHARS = " .`-_':,;=/\"|)\\)ivxclrs{*}I?!][1taeo7zjLunT#JCwfy325Fp6mqSghVd4EgXPGZbYkOA&8U$@KHDBWNMR0QQ"

def lum_to_char(lum, threshhold=0):
    """Map a luminosity value to a character."""
    if lum < threshhold:
        return CHARS[0]
    else:
        return CHARS[int(lum // 2.86)]


def lum_matrix_by_point(image, fontsize):
    """Sample each point at font size dimensions and return matrix of lum values."""
    fnt = ImageFont.truetype('Pillow/Tests/fonts/FreeMono.ttf', fontsize)
    fntsize = fnt.getsize("A")
    return [[image.getpixel((x, y)) for x in range(0, image.width, fntsize[0])]
            for y in range(0, image.height, fntsize[1])]

def reduce_flicker(lum_frames, amt):
    """Lower tolerance for changes in luminosity to reduce flickering between frames."""
    width = len(lum_frames[0][0])
    height = len(lum_frames[0])
    prev = lum_frames[0]
    for i in range(1, len(lum_frames)):
        current = lum_frames[i]
        for y in range(height):
            for x in range(width):
                clum = current[y][x]
                plum = prev[y][x]
                if abs(clum - plum) < amt:
                    clum = plum

        prev = current


def lum_matrix_to_char_matrix(lum_matrix, threshhold=0):
    """Convert luminecience matrix to char matrix."""
    return [[lum_to_char(lum, threshhold) for lum in row] for row in lum_matrix]

def print_chars(char_matrix, image, fontsize):
    """Print a char array over an image."""
    # get a font
    fnt = ImageFont.truetype('Pillow/Tests/fonts/FreeMono.ttf', fontsize)
    y = 0
    fntheight = fnt.getsize("A")[1]

    # get a drawing context
    d = ImageDraw.Draw(image)

    txt_rows = [''.join(row) for row in char_matrix]
    for txt_row in txt_rows:
        # draw text
        d.text((0, y), txt_row, font=fnt, fill=(0, 0, 0, 255))
        y += fntheight



def image_to_ascii(infile_name, outfile_name, font_size):
    """Convert a file to an ascii art file."""
    img = Image.open(infile_name)
    img = img.convert('L')

    outimg = Image.new('RGBA', img.size, (255, 255, 255))

    lums = lum_matrix_by_point(img, font_size)
    char_matrix = lum_matrix_to_char_matrix(lums)

    print_chars(char_matrix, outimg, font_size)

    outimg.save(outfile_name, 'PNG')


def process_sequence(folder_name, font_size, ext='png'):
    """Convert a folder of images to a folder of ascii images."""
    from os import listdir, path, makedirs

    filenames = listdir(folder_name)
    imgs = [name for name in filenames if name[-len(ext):] == ext]
    print "{0} to process".format(len(imgs))

    # Create output folder
    if not path.exists(folder_name + '_out'):
        makedirs(folder_name + '_out')

    print "Opening files."
    imgfiles = [Image.open(folder_name + '/' + img) for img in imgs]

    print "Converting to BnW"
    bwimgs = [img.convert('L') for img in imgfiles]

    print "Converting to lum matrices."
    lum_matrices = [lum_matrix_by_point(bwimg, font_size) for bwimg in bwimgs]

    print "De-jittering lum matrices."
    reduce_flicker(lum_matrices, 20)

    print "Converting to char matrices."
    char_matrices = [lum_matrix_to_char_matrix(lum, 30) for lum in lum_matrices]

    print "Writing imgs"
    for index, img in enumerate(imgs):
        if index % 50 == 0:
            print "on #" + str(index)
        outimg = Image.new('RGBA', bwimgs[index].size, (255, 255, 255))
        print_chars(char_matrices[index], outimg, font_size)

        outpath = folder_name + '_out/' + img
        outimg.save(outpath, 'PNG')


process_sequence('data', 7)

# TODO: Enable print char matrix regradless of font size, use 'threshold' argument
# in lum_matrix_by_point instead of lum_matrix_to_char_matrix, lum_to_char.




ascii/ascii.py
from os import listdir, path, makedirs
from PIL import Image, ImageEnhance, ImageDraw, ImageFont
import click

CHARS = None
with open('chars.txt', 'r') as charstxt:
    for line in charstxt:
        if line[0] == '#':
            continue
        else:
            CHARS = line.rstrip()
            break

if CHARS is None:
    print "chars.txt is messed up..."
    quit()

MIN_CHAR = 0
MAX_CHAR = len(CHARS) - 1



def lum_to_char(lum):
    """Map a luminosity value to a character."""
    # REF: https://stackoverflow.com/questions/345187/math-mapping-numbers
    charidx = int(lum/255.0 * MAX_CHAR)
    return CHARS[charidx]


def lum_matrix_by_point(image, fontsize, font, threshold=0):
    """Sample each point at font size dimensions and return matrix of lum values."""
    fnt = ImageFont.truetype('fonts/' + font + '.ttf', fontsize)
    fntsize = fnt.getsize("A")
    return [[image.getpixel((x, y)) if image.getpixel((x,y)) > threshold else 0
            for x in range(0, image.width, fntsize[0])]
            for y in range(0, image.height, fntsize[1])]

def color_matrix_by_point(image, fontsize, font):
    """Sample each point at font size dimensions and return matrix of lum values."""
    fnt = ImageFont.truetype('fonts/' + font + '.ttf', fontsize)
    fntsize = fnt.getsize("A")
    return [[image.getpixel((x, y)) for x in range(0, image.width, fntsize[0])]
            for y in range(0, image.height, fntsize[1])]

def compress_lum_matrix(lum_matrix, degree):
    return [[i - (i % degree) for i in row] for row in lum_matrix]

def reduce_flicker(lum_frames, amt):
    """Lower tolerance for changes in luminosity to reduce flickering between frames."""
    width = len(lum_frames[0][0])
    height = len(lum_frames[0])
    prev = lum_frames[0]
    for i in range(1, len(lum_frames)):
        current = lum_frames[i]
        for y in range(height):
            for x in range(width):
                clum = current[y][x]
                plum = prev[y][x]

                if abs(clum - plum) < amt:
                    current[y][x] = prev[y][x]

        prev = current



def lum_matrix_to_char_matrix(lum_matrix):
    """Convert luminecience matrix to char matrix."""
    return [[lum_to_char(lum) for lum in row] for row in lum_matrix]

def print_chars(char_matrix, image, fontsize, font, color_matrix=None):
    """Print a char array over an image."""
    # get a font
    fnt = ImageFont.truetype('fonts/' + font + '.ttf', fontsize)
    y = 0

    fntsize = fnt.getsize("A")
    fntheight = fntsize[1]
    fntwidth = fntsize[0]

    imgheight = image.size[1]
    imgwidth = image.size[0]

    x = 0
    y = 0

    # get a drawing context
    d = ImageDraw.Draw(image)

    for rownum, row in enumerate(char_matrix):
        for charnum, char in enumerate(row):
            rgba = (0,0,0,255)
            if color_matrix is not None:
                rgb = color_matrix[rownum][charnum]
                rgba = (rgb[0], rgb[1], rgb[2], 255)

            d.text((x,y), char, font=fnt, fill=rgba)
            x += fntwidth
        x = 0
        y += fntheight



def image_to_ascii(infile_name, outfile_name, font_size):
    """Convert a file to an ascii art file."""
    img = Image.open(infile_name)
    img = img.convert('L')

    outimg = Image.new('RGBA', img.size, (255, 255, 255))

    lums = lum_matrix_by_point(img, font_size, 'arial')
    char_matrix = lum_matrix_to_char_matrix(lums)

    print_chars(char_matrix, outimg, font_size, 'arial')

    outimg.save(outfile_name, 'PNG')


def process_sequence(infile, outfile, font_size, font, reduceflicker, threshold, ext='png', nocolor=False):
    """Convert a folder of images to a folder of ascii images."""

    print "BEEP BEEP BOOP PROCESSING file " + infile

    temp_folder = infile.split('.')[0] + '_temp'

    if outfile is not None:
        pathparts = infile.split('/')[:-1]
        out_folder = '/'.join(pathparts) + '/' + outfile
    else:
        out_folder = infile.split('.')[0] + '_out'

    mkdir(temp_folder)
    mkdir(out_folder)

    print "Converting video to img sequence..."
    run('ffmpeg -i ' + infile + ' ' + temp_folder + '/0%3d.png')

    filenames = listdir(temp_folder)

    imgs = [name for name in filenames if name[-len(ext):] == ext]
    print "{0} frames to process".format(len(imgs))

    print "Opening files."
    imgfiles = [Image.open(temp_folder + '/' + img) for img in imgs]
    if len(imgfiles) == 0:
        print """
        Could not find any images with extension '.{0}' in folder '{1}'
        """.format(ext, temp_folder)
        return


    color_matrices = None
    if not nocolor:
        print "Sampling colors..."
        color_matrices = [color_matrix_by_point(img, font_size, font)
                          for img in imgfiles]


    print "Converting to BnW"
    bwimgs = [img.convert('L') for img in imgfiles]

    print "Converting to lum matrices."
    lum_matrices = [lum_matrix_by_point(bwimg, font_size, font, threshold)
                    for bwimg in bwimgs]

    print "De-jittering lum matrices."
    if reduceflicker > 0:
        lum_matrices = [compress_lum_matrix(matrix, reduceflicker) for matrix in lum_matrices]

    print "Converting to char matrices."
    char_matrices = [lum_matrix_to_char_matrix(lum) for lum in lum_matrices]

    print "Writing imgs"
    for index, img in enumerate(imgs):
        if index % 50 == 0:
            print "on #" + str(index)
        outimg = Image.new('RGBA', bwimgs[index].size, (0, 0, 0, 0))
        color_matrix = (color_matrices[index]
                        if color_matrices is not None else None)
        print_chars(char_matrices[index], outimg, font_size, font, color_matrix)

        outpath = out_folder + '/' + img.split('.')[0] + '.png'
        outimg.save(outpath, 'PNG')

    print "Cleaning up..."
    rmdir(temp_folder)
    # rmdir(out_folder)


def run(command):
    """Execute a bash command."""
    from subprocess import Popen, PIPE
    commands = command.split(' ')
    process = Popen(commands, stdout=PIPE, stderr=PIPE)
    stdout, stderr = process.communicate()
    return (stdout, stderr)

def rmdir(folder_name):
    """Remove a dir."""
    if path.exists(folder_name):
        import shutil
        shutil.rmtree(folder_name)

def mkdir(folder_name):
    """Overwrite dir."""
    rmdir(folder_name)
    makedirs(folder_name)



@click.command()
@click.option('--infile', default='',
help="""
Name of folder containing image sequence to be processed. REQUIRED.
""")
@click.option('--outfile',
help="""
Name of output folder. Default: 'infile + _out'
""")
@click.option('--fontsize', type=int,
help="""
Size of letters. Default: 15
""")
@click.option('--ext',
help="""
Extension of input image files (for distinguishing between input images and
other extraneous files). Default: 'jpg'
""")
@click.option('--reduceflicker', type=int,
help="""
Amount (0-255) by which to reduce rapid flickering between characters. Default: 0
""")
@click.option('--threshold', type=int,
help="""
Luminescience (sp?) threshold. Pixels below this threshold are compressed to 0.
""")
@click.option('--font',
help="""
Name of .ttf file in /fonts to use. Must be a monospaced font, case-sensitive.
(Do not include the extension, just the name).
""")
@click.option('--nocolor', is_flag=True,
help="""
Ignore color.
""")
def process(infile, outfile, fontsize, font, reduceflicker, threshold, ext, nocolor):
    from settings import settings
    if fontsize is None:
        fontsize = settings['fontsize']
    if font is None:
        font = settings['font']
    if reduceflicker is None:
        reduceflicker = settings['reduceflicker']
    if threshold is None:
        threshold = settings['threshold']
    if ext is None:
        ext = settings['ext']

    process_sequence(infile,
                    outfile,
                    fontsize,
                    font,
                    reduceflicker,
                    threshold,
                    ext,
                    nocolor)


if __name__ == '__main__':
    process()
